### DISCUSSION TITLE
Retrieving the name of the file when using s3 Table Function or Engine

### DISCUSSION THREAD

**Discussion by ghost** [ClickHouse/ClickHouse#44849] [Q&A] [answered]
Dear all,

I have a series of similar Parquet files in S3, and I'm trying to create an S3 Table with all the columns from the parquets PLUS one additional column with the name of the file a particular row comes from.

Consider that my S3 directory is something similar to:
- file1.parquet/
-- part1.parquet
-- part2.parquet
-- ...
- file2.parquet/
-- ...

```
SELECT count(*)
FROM s3('https://s3.<my_s3_url>/file**.parquet/*.parquet', 'Parquet', 'col1 String, col2 String, col3 String')
```

I'd like to have as a result:

col1 |  col2 | col3 | filename
foo | bar | baz | file1.parquet
bla | ble | bli | file2.parquet
...

In Redshift, I get this by importing the S3 parquets separately into a staging table, I update it to have the filename, and then insert the data into the final table.

Upon migrating to ClickHouse, I'd like to fully leverage on S3 Table Engine and not having this 2-step import process anymore (for my usecase, I can live with the latency due to using S3 tables). However, I'd still need to have access to the name of the files "on-the-fly".

Anyone can share some hints?

Cheers,

Hugo

**Comment by alexey-milovidov** [comment, marked as answer]
There are two virtual columns available for every s3/file/url table: `_file` and `_path`.

**Reply by ghost** [reply]
Excellent, exactly what I need. However, I'm getting:
` DB::Exception: Column '_path' is not presented in input data.: While executing ParquetBlockInputFormat: While executing S3. (THERE_IS_NO_COLUMN)`

Maybe is it related to https://github.com/ClickHouse/ClickHouse/pull/35586 and https://github.com/ClickHouse/ClickHouse/pull/35834 
 considering that my server is version 22.3.15

I'll update and check how it goes.

Thank you very much!

**Reply by danthegoodman1** [reply]
@alexey-milovidov 
This doesn't seem to work with `url` table function: 
```
SELECT
count(), _file
FROM url('https://datasets-documentation.s3.eu-west-3.amazonaws.com/house_parquet/house_{0,1}.parquet')
GROUP BY
_file
LIMIT 10
```

`Code: 8. DB::Exception: Not found field(_file) in arrow schema:price: int64 not null`
